base/set.jl

```diff
@@ -14,9 +14,14 @@ Set() = Set{Any}()
 function Set{T}(s::KeySet{T, <:Dict{T}}) where {T}
     d = s.dict
     slots = copy(d.slots)
-    keys = copy(d.keys)
-    vals = similar(d.vals, Nothing)
-    _Set(Dict{T,Nothing}(slots, keys, vals, d.ndel, d.count, d.age, d.idxfloor, d.maxprobe))
+    n = length(d.pairs)
+    pairs = Vector{Pair{T,Nothing}}(undef, n)
+    for i in 1:n
+        if isslotfilled(d, i)
+            pairs[i] = d.pairs[i].first::T => nothing
+        end
+    end
+    _Set(Dict{T,Nothing}(slots, pairs, d.ndel, d.count, d.age, d.idxfloor, d.maxprobe))
```

Comment:
This change worries me a little in terms of potential performance regressions. Have you benchmarked this constructor?

Reply:
Unfortunately, you are right. How often is this constructor expected to be used? 

Master
``` julia
typeof(dict) = Dict{Float64, Int64}
typeof(set) = Set{Float64}
0.057843573 s
typeof(dict) = Dict{Int64, Int64}
typeof(set) = Set{Int64}
0.079514255 s
typeof(dict) = Dict{Int32, Int32}
typeof(set) = Set{Int32}
0.031903509 s
typeof(dict) = Dict{Float32, Float32}
typeof(set) = Set{Float32}
0.027489167 s
```

PR
``` julia
typeof(dict) = Dict{Float64, Int64}
typeof(set) = Set{Float64}
0.111745677 s
typeof(dict) = Dict{Int64, Int64}
typeof(set) = Set{Int64}
0.123777739 s
typeof(dict) = Dict{Int32, Int32}
typeof(set) = Set{Int32}
0.088033362 s
typeof(dict) = Dict{Float32, Float32}
typeof(set) = Set{Float32}
0.09261676 s
```

Reply:
You could try playing playing with `@inbounds`, maybe even `@simd`. We can probably also omit the `isslotfilled` for the case where the pair is a bitstype to avoid the extra branch. To be clear: I think this is probably enough of an edge case that we'd take a slight performance hit if there is a benefit for the more common cases, but it would be great if we could minimize the regression here.

Reply:
Should be fixed now, but only if both `{T, V}` types are bitstypes.

Reply:
Can you update the benchmarks? How bad is the regression with e.g. `Dict{Any, Any}` still? Does `@inbounds` help at all?

Reply:
`@inbounds` doesn't have significant impact. Now, the PR seems to be a clear win for this constructor.

Master
``` julia
typeof(dict) = Dict{Int64, Int64}
typeof(set) = Set{Int64}
0.066371097 s
typeof(dict) = Dict{Int64, Any}
typeof(set) = Set{Int64}
1.025274756 s
typeof(dict) = Dict{Any, Int64}
typeof(set) = Set{Any}
1.245573484 s
typeof(dict) = Dict{Any, Any}
typeof(set) = Set{Any}
2.312128789 s
```
PR
``` julia
typeof(dict) = Dict{Int64, Int64}
typeof(set) = Set{Int64}
0.061166678 s
typeof(dict) = Dict{Int64, Any}
typeof(set) = Set{Int64}
0.12504987 s
typeof(dict) = Dict{Any, Int64}
typeof(set) = Set{Any}
0.137963982 s
typeof(dict) = Dict{Any, Any}
typeof(set) = Set{Any}
0.128701097 s
```
Test code
``` julia
module TestDict

using Random
Random.seed!(42)

const n = 10_000_000

for (A,B) in [(Int, Int), (Int, Any), (Any, Int), (Any, Any)]
    T1 = A == Any ? Int : A
    T2 = B == Any ? Int : B
    k = rand(T1, n); k[k.==0] .= typemax(T1)
    values = rand(T2, n); values[values.==0] .= typemax(T2)

    dict = Dict{A, B}()
    @show typeof(dict)

    xn = length(k)
    sizehint!(dict, xn)
    for i in 1:xn
        dict[k[i]] = values[i]
    end

    set = Set(keys(dict))
    @show typeof(set)
    t = @elapsed Set(keys(dict))
    println("$t s")
    
end

end
```

Reply:
For context: this constructor is used in `copy(::KeySet)` (added in #41836).